File: objectModel/TypeScript/__test__/Cdm/ReferencedEntityDeclaration/ReferencedEntityDeclaration.test.ts
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

import { ModelJson } from '../../../Persistence';
import { Model } from '../../../Persistence/ModelJson/types';
import { LocalAdapter } from '../../../Storage';
import {
    CdmCorpusDefinition,
    CdmManifestDefinition,
    CdmTraitReference,
    copyOptions,
    resolveOptions
} from '../../../internal';
import { testHelper } from '../../testHelper';

/**
 * 
 */
describe('Cdm/ReferencedEntityDeclarationTest', () => {
    /**
     * The path between TestDataPath and TestName.
     */
    const testsSubpath: string = 'Cdm/ReferencedEntityDeclaration';

    /**
     * Test that ReferencedEntity is correctly found with path separator as "/" or "\"
     */
    it('TestRefEntityWithSlashPath', async () => {

        const slashCorpus: CdmCorpusDefinition = testHelper.getLocalCorpus(testsSubpath, 'TestRefEntityWithSlashPath');
        const slashLocalPath: string = (slashCorpus.storage.namespaceAdapters.get('local') as LocalAdapter).root;
        var slashAdapter = new LocalAdapterWithSlashPath(slashLocalPath, "/");
        slashCorpus.storage.mount("slash", slashAdapter);
        slashCorpus.storage.defaultNamespace = "slash";

        // load model.json files with paths generated using both '/' and '\'
        var slashManifest = await slashCorpus.fetchObjectAsync<CdmManifestDefinition>('slash:/model.json');

        // manually add the reference model location, path will vary on each machine
        var refModelTrait = slashManifest.exhibitsTraits.item('is.modelConversion.referenceModelMap') as CdmTraitReference;
        var entityPath = slashManifest.entities.allItems[0].entityPath;
        refModelTrait.arguments.allItems[0].value[0].location = slashAdapter.createAdapterPath(entityPath.substring(0, entityPath.lastIndexOf('/')));

        var slashModel = await ModelJson.ManifestPersistence.toData(slashManifest, new resolveOptions(), new copyOptions());

        expect(slashModel)
            .not.toBeUndefined();
        expect(slashModel.entities.length)
            .toBe(1);

        const backSlashCorpus: CdmCorpusDefinition = testHelper.getLocalCorpus(testsSubpath, 'TestRefEntityWithSlashPath');
        var backSlashLocalPath = (backSlashCorpus.storage.namespaceAdapters.get('local') as LocalAdapter).root;
        var backSlashAdapter = new LocalAdapterWithSlashPath(backSlashLocalPath, '\\');
        backSlashCorpus.storage.mount('backslash', backSlashAdapter);
        backSlashCorpus.storage.defaultNamespace = 'backslash';

        var backSlashManifest = await backSlashCorpus.fetchObjectAsync<CdmManifestDefinition>('backslash:/model.json');

        // manually add the reference model location, path will vary on each machine
        var backSlashRefModelTrait = backSlashManifest.exhibitsTraits.item('is.modelConversion.referenceModelMap') as CdmTraitReference;
        var backSlashEntityPath = backSlashManifest.entities.allItems[0].entityPath;
        backSlashRefModelTrait.arguments.allItems[0].value[0].location = backSlashAdapter.createAdapterPath(backSlashEntityPath.substring(0, backSlashEntityPath.lastIndexOf('/')))
            .replace(/\//g, '\\\\');

        const backSlashModel: Model = await ModelJson.ManifestPersistence.toData(backSlashManifest, new resolveOptions(), new copyOptions());

        expect(backSlashModel)
            .not.toBeUndefined();
        expect(backSlashModel.entities.length)
            .toBe(1);
    });

    class LocalAdapterWithSlashPath extends LocalAdapter {
        private separator: string;

        constructor(root: string, separator: string) {
            super(root);
            this.separator = separator;
        }

        public createAdapterPath(corpusPath: string): string {
            const basePath: string = super.createAdapterPath(corpusPath);
            return this.separator == "/" ? basePath.replace(/\\/, '/') : basePath.replace(/\//, '\\');
        }

        public createCorpusPath(adapterPath: string): string {
            return adapterPath;
        }
    }
});

File: objectModel/CSharp/Microsoft.CommonDataModel.ObjectModel.Tests/Cdm/ReferencedEntityDeclaration/ReferencedEntityDeclarationTest.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

namespace Microsoft.CommonDataModel.ObjectModel.Tests.Cdm.ReferencedEntityDeclaration
{
    using Microsoft.CommonDataModel.ObjectModel.Cdm;
    using Microsoft.CommonDataModel.ObjectModel.Persistence.ModelJson;
    using Microsoft.CommonDataModel.ObjectModel.Storage;
    using Microsoft.CommonDataModel.ObjectModel.Utilities;
    using Microsoft.VisualStudio.TestTools.UnitTesting;
    using System.IO;
    using System.Threading.Tasks;

    [TestClass]
    public class ReferencedEntityTest
    {
        /// <summary>
        /// The path between TestDataPath and TestName.
        /// </summary>
        private readonly string testsSubpath = Path.Combine("Cdm", "ReferencedEntityDeclaration");

        /// <summary>
        /// Test that ReferencedEntity is correctly found with path separator as "/" or "\"
        /// </summary>
        [TestMethod]
        public async Task TestRefEntityWithSlashPath()
        {
            var slashCorpus = TestHelper.GetLocalCorpus(testsSubpath, "TestRefEntityWithSlashPath");
            var slashLocalPath = ((LocalAdapter)slashCorpus.Storage.NamespaceAdapters["local"]).Root;
            var slashAdapter = new ModelJsonUnitTestLocalAdapter(slashLocalPath);
            slashCorpus.Storage.Mount("slash", slashAdapter);
            slashCorpus.Storage.DefaultNamespace = "slash";

            // load model.json files with paths generated using both '/' and '\'
            var slashManifest = await slashCorpus.FetchObjectAsync<CdmManifestDefinition>("slash:/model.json");

            // manually add the reference model location, path will vary on each machine
            var refModelTrait = slashManifest.ExhibitsTraits.Item("is.modelConversion.referenceModelMap") as CdmTraitReference;
            var entityPath = slashManifest.Entities[0].EntityPath;
            refModelTrait.Arguments[0].Value[0].location = slashAdapter.CreateAdapterPath(entityPath.Substring(0, entityPath.LastIndexOf("/")));

            var slashModel = await ManifestPersistence.ToData(slashManifest, new ResolveOptions(), new CopyOptions());

            Assert.IsNotNull(slashModel);
            Assert.AreEqual(1, slashModel.Entities.Count);

            var backSlashCorpus = TestHelper.GetLocalCorpus(testsSubpath, "TestRefEntityWithSlashPath");
            var backSlashLocalPath = ((LocalAdapter)backSlashCorpus.Storage.NamespaceAdapters["local"]).Root;
            var backSlashAdapter = new ModelJsonUnitTestLocalAdapter(backSlashLocalPath);
            backSlashCorpus.Storage.Mount("backslash", backSlashAdapter);
            backSlashCorpus.Storage.DefaultNamespace = "backslash";

            var backSlashManifest = await backSlashCorpus.FetchObjectAsync<CdmManifestDefinition>("backslash:/model.json");

            // manually add the reference model location, path will vary on each machine
            var backSlashRefModelTrait = backSlashManifest.ExhibitsTraits.Item("is.modelConversion.referenceModelMap") as CdmTraitReference;
            var backSlashEntityPath = backSlashManifest.Entities[0].EntityPath;
            backSlashRefModelTrait.Arguments[0].Value[0].location = backSlashAdapter.CreateAdapterPath(backSlashEntityPath.Substring(0, backSlashEntityPath.LastIndexOf("/")))
                .Replace("/", "\\\\");

            var backSlashModel = await ManifestPersistence.ToData(backSlashManifest, new ResolveOptions(), new CopyOptions());

            Assert.IsNotNull(backSlashModel);
            Assert.AreEqual(1, backSlashModel.Entities.Count);
        }
    }
}
